jupyter_client/manager.py

```diff
@@ -454,12 +454,21 @@ def is_alive(self):
             # we don't have a kernel
             return False
 
-    def memory_usage(self):
+    @property
+    def kernel_stats(self):
+        """
+        Returns a dict containing functions that gather
+        different stats about the kernel.
+        """
+        return {
+            "memory_usage": self._memory_usage
```

Comment:
This should return the results of `_memory_usage()`, otherwise we'll get into issues for _when_ the consumer tries to get at the memory usage.

`kernel_stats()` should be the method that can be overridden, not `_memory_usage()`.  This way subclasses could even extend the json results with their _provider-specific_ details.

Reply:
@minrk thanks for the review!

I get what you are saying, BUT, few things:

you have said `That approach could be problematic depending on where the consumer of kernel_stats() resides`,  **could you give me an example for such scenario?**

Also, you said, `callers would then have the particular functions in their hands and they can then just call those methods are various intervals? If so, I think it's better to have the callers call the public kernel_stats() method periodically and leave the various functions that produce the results hidden from the caller.`

Yes that was the intent of returning only the function object.
If the json that the function `kernel_stats` returns will grow,
then it means that the time it takes to collect all of the data will increase, meaning that letting it run all of the data collection as soon as it is executed will take more time.

Also, consider that different places that use the function will need different types of data.
Not all of the uses for the `kernel_stats` will use all of the json values, meaning that they are generated for nothing.
So, if for example, lets say we have the need to check every few seconds the ram of the kernel, but we have also added other types of checks into the json, and they all run as soon as you call the function `kernel_stats`, **we will spend getting unneeded data just to get that one value from the json, which will result in a performance decrease.**

I think that letting the user choose the kind of data that he will get from the function is much preferred in terms of performance.

Reply:
@kevin-bates haha sorry for the confusion 😆

>That's okay, but it strikes me as odd since those functions can be called within the kernel_stats() method itself.

I agree, it might also make things complicated 😐

> so the frequency is not very often in the grand scheme of things

This is really depending on the use case.

> If we preserve that goal that none of the stats are expensive, then it should be fine to be able to call this method on 5-10 second intervals w/o detriment

In my opinion it is pretty restricting.
You might have a lot of different types of data gathering added in the future, which will mean that there might come a time where you would need to decide what type of data gathering to leave in and what to take out because all of the functions inside `kernel_stats` will execute and take too much time.

> In addition, other implementations that MUST override kernel_stats() because the kernel is remote, etc. absolutely cannot return functions since those functions do not apply to the caller's context. As a result, the boundary of work must be the public function - kernel_stats().

Yes I agree, let me make myself clear a bit more, I ment that the user should override also the `_memory_usage` function. You wouldn't want to leave the implementation of the function as is and create a new one leaving the old one standing there.

> Btw, when remote access is required to get these stats, the latency of having to go remote in the first place argues that as much information should be gathered at that time, since the cost of the stats call is far greater than the cost to gather the different categories in the first 

I agree.

I thought about it some more, let me propose a new implementation.
It is sort of a combination of the two approaches.
The function `kernel_stats` will receive a list of strings.
Each string representing the type of data that the caller wants to get.
Iterating Over the different types, we will execute the appropriate functions.
This way:
1. The execution and gathering of the data is happening inside the `kernel_stats` function.
2. We execute only the functions that are necessary, this leaves the user much more flexibility because he can specify exactly the data. Because the user specifies the data that is collected, we don't have the restriction of the time for the data gathering functions the `kernel_stats` exposes.
Because of no time limit, we can implement gathering function that take much more time, leaving the user of the function to try and see what time frames are good for him for his use case.
3. Also there will never be a problem another kind of data gathering, since the user is the one responsible for specifying his needs, leaving him responsible for the time it takes to run the data gathering.
4. The user of the function must know what kind of data he is looking for.

I will make a new commit soon.

Reply:
@Hyaxia - this seems like a good approach - thank you.  

I'm still unclear who/what the actual client/caller of this method is and how they ultimately get the data.  I'm assuming this will become exposed in the `api/kernel/{kid}` endpoint somehow or a `api/kernel/{kid}/stats` endpoint is added.   As of now, this method is usable as a form of diagnostics that is added ad-hoc within the NB server code.

I think the list of values to gather will want to be exposed via a config parameter (e.g., `--KernelManager.kernel_stats='memory_stats, latency'`), but that can be added when a formal call to this method is implemented.  In addition, there should be a way to skip this call - perhaps if the config parameter is not defined?   

Since the categories are now specified, their names become more important. Looking at the command line example, it looks like `memory_stats` should become `memory` since _stats_ is redundant.  Or the suffix could be finer grained, `memory_vm`, `memory_rss`, etc. relative to which statistic is being returned.  Just a thought.

Reply:
@kevin-bates thanks!

> Looking at the command line example, it looks like memory_stats should become memory since stats is redundant. Or the suffix could be finer grained, memory_vm, memory_rss, etc. relative to which statistic is being returned

Good point! will change it to a more simpler and shorter version.

> I'm still unclear who/what the actual client/caller of this method is and how they ultimately get the data

Well there might be different use cases.
For example, the first use case that comes to my mind is the one that this whole idea originated from and it is the prometheus one.
Exposing a metric for the prometheus agent to collect data on (this use case is as you said, will be used on the NB server code).
In this case, the caller will be the server side of the notebook, calling the `kernel_stats` function every second or so and getting the memory usage of the kernel which will then be sent to the right metric.
As you said, exposing the method in the API to let an outside service query the data by an HTTP request, mapping the parameters of the HTTP requests to a list of string which is then passed into the function, is a good idea letting it me consumed in additional way.

> I think the list of values to gather will want to be exposed via a config parameter (e.g., --KernelManager.kernel_stats='memory_stats, latency'), but that can be added when a formal call to this method is implemented

How would you implement this kind of configuration? where would you put it?

Reply:
Others can correct me if I'm wrong, but my understanding is that every sub-class of `Configurable` supports the ability to specify runtime options via the command line and/or configuration file.  The command line syntax is `--<className>.<option>=<value>` and config file is `c.<className>.<option>=<value>`.  So for this, the option could be defined on `KernelManager` or `MultiKernelManager` (probably depending on where the `kernel_stats()` method will be called from) and look something akin to [this code in Notebook](https://github.com/jupyter/notebook/blob/master/notebook/notebookapp.py#L1057-L1059).

```python
kernel_statistics = List(Unicode(), config=True,
        help=_("""Comma-separated list of support categories to gather relative to the active kernel.   No statistics are gathered if option is not defined.  Categories are: memory""")
    )
```
You can then specify validation methods to ensure the values are valid.

Reply:
Let me see if I got it right,
So by adding the lines you wrote above, 
`kernel_statistics = List(Unicode(), config=True,
        help=_("""Comma-separated list of support categories to gather relative to the active kernel.   No statistics are gathered if option is not defined.  Categories are: memory""")
    )`

We will then check if there is a configuration and if there is then use it like that code below?
```
if kernel_statistics:
    stats = kernel_statistics
```

If so, how do these configuration tools work together with the `*stats` parameter that the `kernel_stats` receives? wont there be a conflict between those two? I'm not so familiar with the concept 😀

Reply:
I agree.  Logging and _continuing to gather the rest of the keys_ on exceptions is probably the right thing for this since its a background "thread" anyway.  The important thing is to document the expected behavior for errors and unrecognized keys.  You _could_ return a `<key>_errmsg` field containing the applicable exception message, but we should see how this pans out first before going down that road.